InfrastructureManager/agents/azure_agent.py

```diff
@@ -0,0 +1,559 @@
+from azure.mgmt.compute.models import SshConfiguration
+from azure.mgmt.compute.models import SshPublicKey
+from azure.mgmt.compute.models import StorageProfile
+from azure.mgmt.compute.models import VirtualHardDisk
+from azure.mgmt.compute.models import VirtualMachine
+from azure.mgmt.compute.models import VirtualMachineSizeTypes
+
+from azure.mgmt.network import NetworkManagementClient
+from azure.mgmt.network.models import AddressSpace
+from azure.mgmt.network.models import IPAllocationMethod
+from azure.mgmt.network.models import NetworkInterfaceIPConfiguration
+from azure.mgmt.network.models import NetworkInterface
+from azure.mgmt.network.models import PublicIPAddress
+from azure.mgmt.network.models import Subnet
+from azure.mgmt.network.models import VirtualNetwork
+
+from haikunator import Haikunator
+
+# AppScale-specific imports
+from agents.base_agent import AgentConfigurationException
+from agents.base_agent import AgentRuntimeException
+from agents.base_agent import BaseAgent
+from utils import utils
+
+class AzureAgent(BaseAgent):
+  """ AzureAgent defines a specialized BaseAgent that allows for interaction
+  with Microsoft Azure. It authenticates using the ADAL (Active Directory
+  Authentication Library).
+  """
+  # The Azure URL endpoint that receives all the authentication requests.
+  AZURE_AUTH_ENDPOINT = 'https://login.microsoftonline.com/'
+
+  # The Azure Resource URL to get the authentication token using client credentials.
+  AZURE_RESOURCE_URL = 'https://management.core.windows.net/'
+
+  # The default Storage Account name to use for Azure.
+  DEFAULT_STORAGE_ACCT = 'appscalestorage'
+
+  # The default resource group name to use for Azure.
+  DEFAULT_RESOURCE_GROUP = 'appscalegroup'
+
+  # The following constants are string literals that can be used by callers to
+  # index into the parameters that the user passes in, as opposed to having to
+  # type out the strings each time we need them.
+  PARAM_APP_ID = 'app_id'
+  PARAM_APP_SECRET = 'app_secret_key'
+  PARAM_CREDENTIALS = 'credentials'
+  PARAM_EXISTING_RG = 'does_exist'
+  PARAM_GROUP = 'group'
+  PARAM_INSTANCE_IDS = 'instance_ids'
+  PARAM_KEYNAME = 'keyname'
+  PARAM_IMAGE_ID = 'image_id'
+  PARAM_REGION = 'region'
+  PARAM_RESOURCE_GROUP = 'resource_group'
+  PARAM_STORAGE_ACCOUNT = 'storage_account'
+  PARAM_SUBCR_ID = 'subscription_id'
+  PARAM_TENANT_ID = 'tenant_id'
+  PARAM_TEST = 'test'
+  PARAM_TAG = 'group_tag'
+  PARAM_VERBOSE = 'is_verbose'
+  PARAM_ZONE = 'zone'
+
+  # A set that contains all of the items necessary to run AppScale in Azure.
+  REQUIRED_CREDENTIALS = (
+    PARAM_APP_SECRET,
+    PARAM_APP_ID,
+    PARAM_IMAGE_ID,
+    PARAM_KEYNAME,
+    PARAM_SUBCR_ID,
+    PARAM_TENANT_ID,
+    PARAM_ZONE
+  )
+
+  # The admin username needed to create an Azure VM instance.
+  ADMIN_USERNAME = 'azureuser'
+
+  # The file path for the authorized keys on the head node
+  # for an Azure VM.
+  AUTHORIZED_KEYS_FILE = "/home/{}/.ssh/authorized_keys"
```

Comment:
/home is not a good choce to hardcode since the homes directories could be mounted somewhere else. We could use ~ instaed.